plaidcloud/utilities/frame_manager.py

```diff
@@ -1260,137 +1248,147 @@ def apply_rules(df, df_rules, target_columns=['value'], include_once=True, show_
         df['rule_number'] = ''
         df['rule'] = ''
 
-    # Establish new column(s) as blank columns.
+    # Establish new column(s) as blank columns <i>if they do not already exist.</i>
     for column in target_columns:
-        df[column] = ''
+        if column not in df.columns:
+            df[column] = ''
 
     def exclude_matched(include, match):
         """exclude if matched, or if previously excluded"""
-        if match == True:
-            return False
-        else:
-            return include
+        return False if match is True else include
 
-    # rule_num = 0
-    matched_chunks = []
     summary = []
 
-    for index, rule in df_rules.iterrows():
-
+    iterations = list(set(df_rules['iteration']))
+    iterations.sort()
 
-    #for rule in rules:
-        # rule_num = rule_num + 1
-        #rule_condition = rule.get('condition')
-        print('')
-        print('{}.'.format(index))
+    for iteration in iterations:
+        matched_chunks = []
+        df['include'] = True
 
-        # Find subset based on condition
-        input_length = len(df[df['include'] == True])
-        if rule['condition'] is not None and rule['condition'] != '' and str(rule['condition']) != 'nan':
-            try:
-                df_subset = df[df['include'] == True].query(rule['condition'], engine='python')
-                print('{} - input length'.format(input_length))
-                if show_rules == True:
-                    df_subset['rule_number'] = str(index)
-                    df_subset['rule'] = str(rule['condition'])
-            except Exception as e:
-                df_subset = pd.DataFrame()
-                def add_message(log):
-                    return '<{} ::: {}>'.format(e, log)  # removed redundant rule['condition'] param from format string
-                if show_rules == True:
-                    df['log'] = list(map(add_message, df['log']))
-                error_msg = ' (rule_num {0}) {1} error: {2}'.format(index, rule['condition'], e)
-                logger.exception('EXCEPTION {}'.format(error_msg))
-        else:
-            df_subset = df[df['include'] == True]
-
-        # Populate target columns as specified in split
-        for column in target_columns:
-            df_subset[column] = rule[column]
-
-        # need to find a way to flip the flag once data has been selected
+        if iteration == '2' or iteration == '3':
+            a = 1
```

Comment:
Probably shouldn't be in final code

Reply:
Is iteration a string? If you had 10+ iterations, would that sort numerically, or alpha-numerically?